README.md
# openai-cli

This is a simple, command-line wrapper for the OpenAI API to allow for a quicker, more flowing interaction with OpenAI's natural language models.

Not all of the API is currently working via the CLI, but I am the process of adding more commands as time goes on. Please reach out with an issue or pull request if you'd like a certain API function to be developed quicker.

Built on: [openai-node](https://github.com/openai/openai-node)

## ChatGPT

The ChatGPT API was finally released today! (3/1/2023)

We've gone ahead and implemented it here. **The logging system may not be completely up to date or functional, but it will be at some point in the future.**

Simply use the `--chat` argument, and supply a `-p` prompt at the start to set the [personality or the 'assistant' info of the bot](https://platform.openai.com/docs/guides/chat/introduction). You can also use the `--name` argument to give the bot a name, purely for cosmetics. Streaming may not work as well, as the chat functionality is compartmentalized within the app.

Chat has only one special command so far, `!CLEAR`, which clears the console. More chat specific commands will come later or upon request.

## Warning

_Taken directly from the [openai-node](https://github.com/openai/openai-node) repo:_

**Important note:** This library [exposes endpoints that] is meant for server-side usage only, as using it in client-side browser code will expose your secret API key. See here for more details.

## Installation

1. Fork/Clone this repository.
2. Run `yarn install` within this repo.
3. Create a `.env` file, verify that it contains a valid `OPENAI_API_KEY` field using your private API key.
4. Begin utilizing the commands below.

## Flags

| Flag                        | Description                                                                                                                                                                                                             | Default Value                   |
| --------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------- |
| -b, --beautify              | Formats the output from the request to look beautiful. Note: 'p' was already taken by 'prompt', let alone 'prettier'.                                                                                                   | false                           |
| -c, --config [string]       | Import specified config file to use with query to endpoint. Must be in JSON format. Config values match the flags listed in this table. Config files WILL ALWAYS OVERWRITE command-line args.                           |                                 |
| --chat                      | Enables ChatGPT related API requests..                                                                                                                                                                                  | false                           |
| -co, --count [integer]      | Specifies the number of images that should be produced using DALL-E.                                                                                                                                                    | 1                               |
| -d, --dalle [string]        | Enables generation of DALL-E images using the command-line with a specified prompt.                                                                                                                                     |                                 |
| -ft, --fine_tune [string]   | Train and upload a fine-tune dataset into OpenAI's servers to create a new model.                                                                                                                                       |                                 |
| -h, --help                  | Display help for all commands listed.                                                                                                                                                                                   |                                 |
| -i, --input [string]        | CURRENTLY BROKEN. DO NOT USE. Import external prompt or multiple prompts from local JSON file.                                                                                                                          |                                 |
| -ij, --inject [string]      | Useful when using an external file for prompts, can inject a string (append) to the end of a prompt.                                                                                                                    |                                 |
| -s, --size [string]         | The preferred resolution of images produced by DALL-E. (256x256, 512x512, 1024x1024)                                                                                                                                    | 256x256                         |
| -m, --model [string]        | A string value representing the model name hosted on OpenAI's servers.                                                                                                                                                  | 'text-davinci-003'              |
| -mt, --max_tokens [integer] | Changes number of max tokens output by the current model. This directly correlates with API usage and pricing.                                                                                                          | 64                              |
| --name [string]             | The name of the chat bot you wish to use when using ChatGPT. Purely cosmetic in the terminal.                                                                                                                           | 'Bot'                           |
| -o, --output [string]       | Redirects the generated log file to a new output location. Note that if the file already exists, it will be overwritten without warning.                                                                                | 'logs/output/[date].output.log' |
| -p, --prompt [string]       | Passes a string value directly into model as a prompt.                                                                                                                                                                  |                                 |
| -s, --size [string]         | The image size of the attempted generated image. Only approved resolutions will be accepted. Must use a DALL-E model. (256x256, 512x512, 1024x1024)                                                                     | "256x256"                       |
| -st, --stream               | If enabled, will stream data directly from the OpenAI endpoint to your terminal. Warning: this may sometimes not work as intended, as this is a workaround.                                                             | false                           |
| -su, --suffix               | Can append more context to the end of a prompt. Suffix cannot be used while echo is being used as well. [Link to suffix docs.](https://beta.openai.com/docs/api-reference/completions/create#completions/create-suffix) |                                 |
| -t, --temp [float]          | Modified generation temperature. Temperature approaching 1 equates to more randomness in output. Value from 0 to 1.                                                                                                     | 0.7                             |
| -v, --verbose               | Enables extraneous debugging information for most functions and services that support this flag.                                                                                                                        | false                           |

## Command-Line Example

```txt
$ node ./src/index.js -bv -t 1 -mt 256 -p "What are 5 main security concepts when designing an API?"

info: {"version":0.2,"beautify_enabled":true,"temperature":1,"model":"text-davinci-003","max_tokens":256}
info:
The five main security concepts when designing an API include:

1. Authentication: API consumers must be authenticated in order to access the API.

2. Authorization: API consumers must be authorized to access the API.

3. Encryption: API communications should be encrypted in order to ensure privacy.

4. Rate limiting: API consumers should be rate-limited in order to prevent malicious or excessive use.

5. Input validation: API consumers should be required to provide valid input in order to prevent security risks.

Done in 4.61s.
```

### Output

By default, all generated output will be directed into a folder labeled `logs/output/[date].output.log`. This is to save working ideas or review generated content.


CHANGELOG.md
## Version 5

-   Updated `openai` library to `1.22.19`, up from `1.22.15`.
-   Added `inquirer` dependency for chat interaction.
-   Added `--chat` flag to allow access to the ChatGPT API endpoint! (Finally!!! 🥳)
-   Added `--name` flag to allow users to set the name of the chat bot in the terminal.

### Verison 4

-   Added `-fp, --freq_pen` flag to allow modifications to the frequency penalty.
-   Added `-pp, --pres_pen` flag to allow modifications to the presence penalty.
-   Added `-ij, --inject` flag to allow appending text to the end of a prompt string. Great for using external/config prompts.
-   Added `-n, --number` flag to edit the amount of completions returned by the endpoint.
-   Added `-bo, --best_of` flag to generate the completions server-side and returns the "best" result.
-   Added `-l, --log_probs` flag to include the log probabilities on the most likely tokens.
-   Added `-tp, --top_p` flag to give an alternative to sampling output.
-   Added `-su, --suffix` flag to continue migrating the API. Suffixes are not the same as injections, please refer to the documentation in the README.md.
-   Added `--stop` flag to allow up to 4 stop sequences.
-   Added `-r, --raw` flag to allow output of entire axios response object.
-   Tweaking `--stream` flag performance and appearance.
-   Updated verbose object (vObj) to use shorter descriptions.
-   Fixed issued with config files not actually assuming new parameters when using it.
-   Streamlined core functions.

### Version 3

-   Added `-st, --stream` flag to enable response streaming from the API endpoint, similar to how the playground functions.
-   Added `-c, --config` flag to import model settings from an external JSON file.
-   Added `-e, --echo` flag to give the option to echo the prompt alongside the completed output.
-   Updated `--temperature` to `--temp` to make it easier to type.
-   Updated `-n, --number` to `-co, --count` for DALL-E endpoints to make more sense.
-   Updated the `-v, --verbose` flag to now include verbose information in generated logs while beautify is enabled, and only output completed time and info tags when using this flag.
-   Changed default `temp` value from 1 to 0.7, reflecting OpenAI's default temperature.
-   Updated README.md with more general and command information, fixed typos and grammar, updated URLs.

### Version 2

-   Implemented DALL-E API.
-   Setting up the foundations for using external input as prompts. (Not yet completely set up yet!)
-   General README fixes, updates to the command table.

### Version 1

-   Initial setup! Beta Release!
-   Basic OpenAI API endpoint capabilities.
-   Setting groundwork for more complex API / CLI features.


src/index.js
const fs = require('fs');
const util = require('util');

const { options, logger, winstonAddFileTransport } = require('../lib/utils');
require('dotenv').config();

const { Configuration, OpenAIApi } = require('openai');
const winston = require('winston');
const configuration = new Configuration({
	apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

/**
 * Both submits and trains the fine-tune dataset for the specified model.
 * Note: This does not verify if the training set is valid. Please refer to OpenAI's documentation
 *       on validating datasets.
 * @param {string} payload - The location of the pre-verified training set.
 * @param {string} model - The required model that the dataset will fine-tune against.
 * @param {string} suffix - String value applied to the model name for faster locating.
 * @param {boolean} verbose - Outputs the response object from the completed fine-tune job.
 */
const trainFineTuneModel = async (
	payload,
	model = 'davinci',
	suffix = 'openai-cli',
	verbose = options.verbose
) => {
	// ? Upload file to OpenAI servers.
	if (payload !== null || payload !== '') {
		let uploadFile;

		// Attempt to upload the training set.
		try {
			if (payload.includes('.jsonl')) {
				console.log('Inside!');
				uploadFile = await openai.createFile(
					fs.createReadStream(payload, { encoding: 'utf-8' }),
					'fine-tune'
				);
			} else {
				logger.error(
					'Training dataset is not of type .jsonl or provide the correct extension.'
				);
				return;
			}
			// Train the fine-tune model against the data.
			try {
				const fineTuneResponse = await openai.createFineTune({
					training_file: uploadFile.data.id,
					model: `${model}`,
					suffix: `${suffix}`,
				});
				logger.info(
					'Submitted! Please give it time to process on the OpenAI servers...'
				);
				verbose ? console.log(fineTuneResponse) : null;
			} catch (e) {
				logger.error(
					`Error submitting your fine-tune model. Please try again.\n${e}`
				);
			}
			return;
		} catch (e) {
			logger.error(`Unable to load file from ${payload}.\n${e}`);
		}
	}
};

// ? Separating logic to expand on this later. Keep it simple for now.
const beautify = (string) => `\n\n${string.trim()}\n`.replace(/\\n/g, '\n');

const main = async () => {
	//
	// If we're not in production then log to the `console` with the format:
	// `${info.level}: ${info.message} JSON.stringify({ ...rest }) `
	//
	if (process.env.NODE_ENV !== 'production' && !options.chat) {
		logger.add(
			new winston.transports.Console({
				format: winston.format.simple(),
			})
		);
	}

	// ? Config file overwrites command-line arguments!
	// ? Multiple prompt query must be used with the -i flag.
	// TODO: Organize later.
	if (options.config) {
		const config = JSON.parse(
			fs.readFileSync(options.config, { encoding: 'utf-8' })
		);
		if (config.temp) options.temp = config.temp;
		if (config.max_tokens) options.max_tokens = config.max_tokens;
		if (config.beautify) options.beautify = config.beautify;
		if (config.model) options.model = config.model;
		if (config.top_p) options.top_p = config.top_p;
		if (config.n) options.n = config.n;
		if (config.dalle) options.dalle = config.dalle;
		if (config.size) options.size = config.size;
		if (config.count) options.count = config.count;
		if (config.raw) options.raw = config.raw;
		if (config.output) options.output = config.output;
		if (config.prompt) options.prompt = config.prompt;
		if (config.stream) options.stream = config.stream;
		if (config.pres_pen) options.pres_pen = config.pres_pen;
		if (config.freq_pen) options.freq_pen = config.freq_pen;
		if (config.verbose) options.verbose = config.verbose;
		if (config.input) options.input = config.input;
		if (config.chat) options.chat = config.chat;
		if (config.name) options.name = config.name;
	}

	// Verbose Object (vObj) MUST be initialized BEFORE config updates above ^
	// If there is no default value in .\lib\utils, and is not supplied, it will not appear in vObj.
	const vObj = {
		version: 0.4,
		beautify: options.beautify,
		temp: parseFloat(options.temp),
		prompt: options.prompt,
		n: options.n,
		injection: options.inject,
		stream: options.stream,
		echo: options.echo,
		suffix: options.suffix,
		top_p: parseFloat(options.top_p),
		n: options.n,
		frequency_penalty: parseFloat(options.freq_pen),
		presence_penalty: parseFloat(options.pres_pen),
		model: options.prompt ? `${options.model}` : `DALL-E MODE`,
		chat: options.chat,
		name: options.name,
		max_tokens: options.prompt
			? parseInt(options.max_tokens)
			: `DALL-E MODE`,
		stop: options.stop,
		raw: options.raw,
		dalle: options.dalle,
		count: options.dalle ? options.count : `GPT MODE`,
		size: options.dalle ? options.size : `GPT MODE`,
	};

	// Does this have to be last?
	const completionRequest = {
		model: `${options.model}`,
		prompt: `${
			options.prompt + (options.inject ? ' ' + options.inject : '')
		}`,
		suffix: options.suffix,
		temperature: parseFloat(options.temp),
		max_tokens: parseInt(options.max_tokens),
		echo: options.echo,
		frequency_penalty: parseFloat(options.freq_pen),
		presence_penalty: parseFloat(options.pres_pen),
		suffix: options.suffix,
		stream: options.stream,
		top_p: parseFloat(options.top_p),
		n: options.n,
		best_of: options.best_of,
		logprobs: parseInt(options.log_probs),
		stop: typeof options.stop == [] ? options.stop.flat() : options.stop,
	};

	// Please refer to ./lib/utils for more information on arguments.
	if (options.output) {
		winstonAddFileTransport(
			options.output,
			options.verbose ? JSON.stringify(vObj) : ''
		);
	}

	if (options.raw) {
		// Returns the ENTIRE response object.
		const response = await openai.createCompletion(completionRequest);
		options.verbose ? logger.info(JSON.stringify(vObj)) : null;
		// logger.info(JSON.stringify(response));
		logger.info(
			util.inspect(response, {
				showHidden: false,
				depth: null,
				colors: true,
			})
		);
		return;
	}

	// External prompt input
	// May have some issues!
	if (options.input) {
		const input = fs.readFileSync(options.input, { encoding: 'utf-8' });
		const jsonData = JSON.parse(input);

		for (let key in jsonData) {
			const response = await openai.createCompletion(completionRequest, {
				responseType: 'text',
			});

			// options.log_probs ? logger.info(response.data.choices) : 'Log probs not enabled?';

			options.beautify
				? logger.info(beautify(response.data.cshoices[0].text))
				: logger.info(JSON.stringify(response.data.choices));
		}
	}

	// Streaming Workaround! (May get fixed soon?)
	// Streaming is currently handled differently than the general prompt request.
	// This requires handling its own output and any other quirks that provides.
	if (options.stream && options.prompt && !options.chat) {
		// https://2ality.com/2018/04/async-iter-nodejs.html#generator-%231%3A-from-chunks-to-lines
		async function* chunksToLines(chunksAsync) {
			let previous = '';
			for await (const chunk of chunksAsync) {
				const bufferChunk = Buffer.isBuffer(chunk)
					? chunk
					: Buffer.from(chunk);
				previous += bufferChunk;
				let eolIndex;
				while ((eolIndex = previous.indexOf('\n')) >= 0) {
					// line includes the EOL
					const line = previous.slice(0, eolIndex + 1).trimEnd();
					if (line === 'data: [DONE]') break;
					if (line.startsWith('data: ')) yield line;
					previous = previous.slice(eolIndex + 1);
				}
			}
		}

		async function* linesToMessages(linesAsync) {
			for await (const line of linesAsync) {
				const message = line.substring('data :'.length);

				yield message;
			}
		}

		async function* streamCompletion(data) {
			yield* linesToMessages(chunksToLines(data));
		}

		let output = '';

		try {
			options.verbose ? logger.info(JSON.stringify(vObj)) : null;

			const completion = await openai.createCompletion(
				completionRequest,
				{ responseType: 'stream' }
			);

			for await (const message of streamCompletion(completion.data)) {
				try {
					const parsed = JSON.parse(message);
					const { text } = parsed.choices[0];

					process.stdout.write(text);
					output += text;
				} catch (error) {
					logger.error(
						'Could not JSON parse stream message',
						message,
						error
					);
				}
			}
			process.stdout.write('\n\n');

			// Manually append to file, creating a new logger would complicate things.
			fs.appendFileSync(
				`logs/output/${new Date()
					.toISOString()
					.replace(/T.*/, '')
					.split('-')
					.join('-')}.output.log`,
				`${JSON.stringify(vObj)}\n${output}`,
				(err) => {
					if (err) throw err;
				}
			);
		} catch (error) {
			if (error.response?.status) {
				logger.error(error.response.status, error.message);

				for await (const data of error.response.data) {
					const message = data.toString();

					try {
						const parsed = JSON.parse(message);

						logger.error(
							'An error occurred during OpenAI request: ',
							parsed
						);
					} catch (error) {
						logger.error(
							'An error occurred during OpenAI request: ',
							message
						);
					}
				}
			} else {
				logger.error('An error occurred during OpenAI request', error);
			}
		}
	} else if (!options.chat) {
		const response = await openai.createCompletion(completionRequest);
		options.verbose ? logger.info(JSON.stringify(vObj)) : null;

		options.beautify
			? logger.info(beautify(response.data.choices[0].text))
			: logger.info(JSON.stringify(response.data.choices));
	}

	// DALLE Image Generation
	if (options.dalle) {
		const response = await openai.createImage({
			prompt: `${options.dalle}`,
			n: parseInt(options.count),
			size: options.size,
		});
		if (response.data.data.length > 1) {
			response.data.data.forEach((image, i) => {
				logger.info(
					beautify(`Link to image (#${i + 1}):\t${image.url}`)
				);
			});
		} else {
			logger.info(
				beautify(`Link to image:\t${response.data.data[0].url}`)
			);
		}
	}

	if (options.chat) {
		// Must import dynamically
		(async () => {
			const inquirerPromise = import('inquirer');
			const inquirer = await inquirerPromise;

			const chatCommands = ['!CLEAR'];

			// Set message start and store for later
			let history = [
				{
					role: 'assistant',
					content: options.prompt || '',
				},
			];

			options.verbose ? console.log(JSON.stringify(vObj)) : null;

			const messageLoop = () => {
				// Create message loop
				inquirer.default
					.prompt({
						type: 'input',
						name: 'userinput',
						message: 'User:',
						prefix: '\n🔎',
					})
					.then(async (answers) => {
						switch (answers.userinput) {
							// Custom commands.
							case '!CLEAR':
								console.clear();
								break;
							default:
								// Harded coded model for now, no other models present.
								const response =
									await openai.createChatCompletion({
										model: 'gpt-3.5-turbo',
										messages: [
											...history,
											{
												role: 'user',
												content: answers.userinput,
											},
										],
										frequency_penalty: parseFloat(
											options.freq_pen
										),
										temperature: parseFloat(options.temp),
										top_p: parseFloat(options.top_p),
									});

								// Add user response and assistant responses to history.
								history.push({
									role: 'user',
									content: answers.userinput,
								});

								history.push({
									role: 'assistant',
									content:
										response.data.choices[0].message
											.content,
								});

								// Output
								console.log(
									`\n🧠 ${options.name}: ${beautify(
										JSON.stringify(
											response.data.choices.at(-1).message
												.content
										)
									)
										.replace(/^"+|"+$/g, '')
										.replace(/^[\s\n]+|[\s\n]+$/g, '')
										.replace(/^"|"$/g, '')
										.replace(/\\/g, '')}`
								);
								// Log the inputs and outputs.
								logger.info(JSON.stringify(answers.userinput));

								logger.info(
									JSON.stringify(
										response.data.choices.at(-1).message
											.content
									)
								);
								break;
						}
						messageLoop();
					});
			};
			messageLoop();
		})();
	}

	// ? Remember! Fine-tuning can be pricey!
	// ? https://openai.com/api/pricing/
	if (options.fine_tune) {
		trainFineTuneModel(options.fine_tune);
	}
};

main();


lib/utils.js
const commander = require('commander');
const winston = require('winston');

// TODO: These need to be alphabatized!

commander
	.name('openai-cli')
	.version('0.4', '--version', 'Outputs the current verison number.')
	.description(
		'A simple, command-line Node.js wrapper for OpenAI built in JavaScript. Please visit the Github repository: https://github.com/justin-carver/openai-cli for more information on contributing and opening issues.'
	)
	.usage('[OPTIONS]...')
	.option(
		'-b, --beautify',
		"Formats the HTTP response from OpenAI's API to look beautiful. Note: 'p' was already taken by 'prompt', let alone 'prettier'.",
		false
	)
	.option(
		'-bo --best_of <integer>',
		'Generates best_of completions server-side and returns the "best" (the one with the highest log probability per token). Results cannot be streamed. When used with n, best_of controls the number of candidate completions and n specifies how many to return – best_of must be greater than n. Note: Because this parameter generates many completions, it can quickly consume your token quota. Use carefully and ensure that you have reasonable settings for max_tokens and stop. Cannot be used while streaming.',
		1
	)
	.option(
		'-c, --config <string>',
		'Relative path to a config.json file that can be imported for settings.'
	)
	.option('--chat', 'Enables ChatGPT related API requests.')
	.option(
		'-d, --dalle <string>',
		'Queries the DALL-E endpoint for image generation using the supplied prompt.'
	)
	.option(
		'-e, --echo',
		'If echo is enabled, the prompt will be echoed back with the completion, regardless if command is verbose.'
	)
	.option(
		'-fp, --freq_pen <float>',
		"The frequency penalty. Number between -2.0 and 2.0. Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim.",
		parseFloat(0)
	)
	.option(
		'-l, --log_probs <integer>',
		'Include the log probabilities on the logprobs most likely tokens, as well the chosen tokens. For example, if logprobs is 5, the API will return a list of the 5 most likely tokens. The API will always return the logprob of the sampled token, so there may be up to logprobs+1 elements in the response.'
	)
	.option(
		'-o, --output <string>',
		'Pipe the completion output into a non-formatted specific file. All entries are appended, not overwritten. File name is rotated daily.',
		`logs/output/${new Date()
			.toISOString()
			.replace(/T.*/, '')
			.split('-')
			.join('-')}.output.log`
	)
	.option(
		'-pp, --pres_pen <float>',
		"The presence penalty. Number between -2.0 and 2.0. Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics.",
		parseFloat(0)
	)
	.option(
		'-i, --input <string>',
		'CURRENTLY BROKEN. DO NOT USE. Takes an input file and uses this as the prompt. The input MUST be in the .jsonl format.'
	)
	.option(
		'-ij, --inject <string>',
		'Useful when using an external file for prompts, can inject a string (append) to the end of a prompt.'
	)
	.option(
		'-m, --model <string>',
		"A string value representing a specified model name hosted on OpenAI's servers.",
		'text-davinci-003'
	)
	.option(
		'-n, --number <integer>',
		'How many completions will be generated for each request.',
		1
	)
	.option(
		'--name <string>',
		'The name of the chat bot you wish to use when using ChatGPT. Purely cosmetic in the terminal.',
		'Bot'
	)
	.option(
		'-p, --prompt <string>',
		'Passes a string value directly into the model as a prompt and uses the OpenAI completion endpoint.'
	)
	.option(
		'-t, --temp <float>',
		'Modified generation temperature. Temperature approaching 1 equates to more randomness in output. Value from 0 to 1.',
		parseFloat(0.7)
	)
	.option(
		'-tp, --top_p <float>',
		'An alternative to sampling with temperature, called nucleus sampling, where the model considers the results of the tokens with top_p probability mass. So 0.1 means only the tokens comprising the top 10% probability mass are considered.',
		1
	)
	.option(
		'-v, --verbose',
		'Enables extraneous debugging information for most functions and services that support this flag.',
		false
	)
	.option(
		'-mt, --max_tokens <integer>',
		'Changes number of max_tokens output by the current model. This directly correlates with API usage and pricing.',
		64
	)
	.option(
		'-ft, --fine_tune <string>',
		"Train and upload a fine-tune dataset into OpenAI's servers to create a new model."
	)
	.option(
		'-co, --count <integer>',
		'The number of images to generate using the DALL-E endpoint. Only works for this API.',
		1
	)
	.option(
		'-s, --size <string>',
		'The image size of the attempted generated image. Only approved resolutions will be accepted. (256x256, 512x512, 1024x1024)',
		'256x256'
	)
	.option(
		'-st, --stream',
		'If enabled, will stream data directly from the OpenAI endpoint to your browser. Warning: this may not work as intended, as this is a workaround.',
		false
	)
	.option(
		'--stop <sequences...>',
		'Up to 4 sequences where the API will stop generating further tokens. The returned text will not contain the stop sequence.',
		(...strings) => strings
	)
	.option(
		'-su, --suffix <string>',
		'Can append more information to the end of a prompt. Suffix cannot be used while echo is being used as well.'
	)
	.option(
		'-r, --raw',
		'Will output the ENTIRE raw response object to the console. Does not work with beautify flag. Enabling raw response while streaming will result in very large ouput!',
		false
	)
	.showSuggestionAfterError(true)
	.showHelpAfterError(
		'\n(type --help for additional information on commands)\n'
	)
	.configureHelp({
		sortSubcommands: true,
		sortOptions: true,
	})
	.addHelpText(
		'after',
		'\nUsage: \nyarn dev -bv -t 1 -mt 256 -p "What are 5 main security concepts when designing an API?"'
	)
	.addHelpText(
		'after',
		'\nPlease view the CHANGELOG.md file for more information on what has changed from the previous version.'
	)
	.addHelpText(
		'after',
		"\n💡 Remember to follow OpenAI's usage guidelines.\nhttps://beta.openai.com/docs/usage-policies\n"
	)
	.parse(process.argv);

const logger = winston.createLogger({
	level: process.env.LOG_LEVEL || 'info',
	format: winston.format.json(),
	transports: [
		//
		// - Write all logs with importance level of `error` or less to `error.log`
		// - Write all logs with importance level of `info` or less to `combined.log`
		//
		new winston.transports.File({
			filename: 'logs/system/error.log',
			level: 'error',
		}),
		new winston.transports.File({ filename: 'logs/system/combined.log' }),
	],
});

exports.winstonAddFileTransport = (filename, verboseContent) => {
	logger.add(
		new winston.transports.File({
			filename: `${filename}`,
			format: winston.format.printf(
				(info) => `${verboseContent ? verboseContent + '\n' : ''}`
			),
		})
	);
};
exports.logger = logger;
exports.options = commander.opts();
